The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3383528762/jobs/5619510407) [(plain)](https://github.com/rust-lang/rust/commit/f4d3ba5aab056278aa3f3642dae0f2c1654e53ed/checks/9266129820/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
    Checking rustc_codegen_ssa v0.0.0 (/checkout/compiler/rustc_codegen_ssa)
    Checking rustc_mir_transform v0.0.0 (/checkout/compiler/rustc_mir_transform)
    Checking rustc_borrowck v0.0.0 (/checkout/compiler/rustc_borrowck)
error[E0405]: cannot find trait `AddToDiagnostic` in this scope
    |
    |
630 |         cb: impl Fn(Span) -> impl AddToDiagnostic,
    |
help: consider importing this trait
    |
3   | use rustc_errors::AddToDiagnostic;
3   | use rustc_errors::AddToDiagnostic;
    |

error[E0425]: cannot find function `f` in this scope
   --> compiler/rustc_borrowck/src/diagnostics/mod.rs:635:5
    |
635 |                 f(capture_kind_span);
    |
help: consider importing one of these items
    |
3   | use crate::diagnostics::sym::f;
---
    |
3   | use rustc_span::sym::f;
    |

error[E0562]: `impl Trait` only allowed in function and inherent method return types, not in `Fn` trait return
    |
    |
630 |         cb: impl Fn(Span) -> impl AddToDiagnostic,


error[E0599]: no method named `var_span_subdiag_label` found for enum `UseSpans` in the current scope
    |
    |
732 |         borrow_spans.var_span_subdiag_label(
    |                      ^^^^^^^^^^^^^^^^^^^^^^ help: there is a method with a similar name: `var_span_label`
   ::: compiler/rustc_borrowck/src/diagnostics/mod.rs:513:1
    |
    |
513 | pub(super) enum UseSpans<'tcx> {
    | ------------------------------ method `var_span_subdiag_label` not found for this enum

error[E0559]: variant `CaptureVarKind::Here` has no field named `capture_kind_span`
    |
639 |                     capture_kind_span,
639 |                     capture_kind_span,
    |                     ^^^^^^^^^^^^^^^^^ `CaptureVarKind::Here` does not have this field
    |
    = note: available fields are: `kind_desc`, `kind_span`
Some errors have detailed explanations: E0405, E0425, E0559, E0562, E0599.
For more information about an error, try `rustc --explain E0405`.
error: could not compile `rustc_borrowck` due to 6 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

--- a/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
@@ -724,13 +724,19 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
             borrow_span,
             &self.describe_any_place(borrow.borrowed_place.as_ref()),
         );
-
-        borrow_spans.var_span_label(
+        borrow_spans.var_subdiag(
             &mut err,
-            {
+            |var_span| {
+                use crate::session_diagnostics::CaptureVarCause::*;
                 let place = &borrow.borrowed_place;
                 let desc_place = self.describe_any_place(place.as_ref());
-                format!("borrow occurs due to use of {}{}", desc_place, borrow_spans.describe())
+                match borrow_spans {
+                    UseSpans::ClosureUse { generator_kind, .. } => match generator_kind {
+                        Some(_) => BorrowUsePlaceGenerator { place: desc_place, var_span },
+                        None => BorrowUsePlaceClosure { place: desc_place, var_span },
+                    },
+                    _ => BorrowUsePlace { place: desc_place, var_span },
+                }
             },
             "mutable",
         );

--- a/compiler/rustc_borrowck/src/diagnostics/mod.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/mod.rs
@@ -623,6 +623,26 @@ impl UseSpans<'_> {
         }
     }
 
+    /// Add a subdiagnostic to the use of the captured variable, if it exists.
+    pub(super) fn var_subdiag(
+        self,
+        err: &mut Diagnostic,
+        f: impl Fn(Span) -> crate::session_diagnostics::CaptureVarCause,
+        kind_desc: impl Into<String>,
+    ) {
+        if let UseSpans::ClosureUse { capture_kind_span, path_span, .. } = self {
+            if capture_kind_span == path_span {
+                err.subdiagnostic(f(capture_kind_span));
+            } else {
+                err.subdiagnostic(crate::session_diagnostics::CaptureVarKind {
+                    kind_desc: kind_desc.into(),
+                    kind_span: capture_kind_span,
+                });
+                err.subdiagnostic(f(path_span));
+            }
+        }
+    }
+
     /// Returns `false` if this place is not used in a closure.
     pub(super) fn for_closure(&self) -> bool {
         match *self {

--- a/compiler/rustc_borrowck/src/session_diagnostics.rs
+++ b/compiler/rustc_borrowck/src/session_diagnostics.rs
@@ -148,3 +148,33 @@ pub(crate) enum RequireStaticErr {
         multi_span: MultiSpan,
     },
 }
+
+#[derive(Subdiagnostic)]
+#[label(borrowck_capture_kind_label)]
+pub(crate) struct CaptureVarKind {
+    pub kind_desc: String,
+    #[primary_span]
+    pub kind_span: Span,
+}
+
+#[derive(Subdiagnostic)]
+pub(crate) enum CaptureVarCause {
+    #[label(borrowck_var_borrow_by_use_place)]
+    BorrowUsePlace {
+        place: String,
+        #[primary_span]
+        var_span: Span,
+    },
+    #[label(borrowck_var_borrow_by_use_place_in_generator)]
+    BorrowUsePlaceGenerator {
+        place: String,
+        #[primary_span]
+        var_span: Span,
+    },
+    #[label(borrowck_var_borrow_by_use_place_in_closure)]
+    BorrowUsePlaceClosure {
+        place: String,
+        #[primary_span]
+        var_span: Span,
+    },
+}

--- a/compiler/rustc_error_messages/locales/en-US/borrowck.ftl
+++ b/compiler/rustc_error_messages/locales/en-US/borrowck.ftl
@@ -58,3 +58,15 @@ borrowck_returned_lifetime_short =
 
 borrowck_used_impl_require_static =
     the used `impl` has a `'static` requirement
+
+borrowck_capture_kind_label =
+    capture is {$kind_desc} because of use here
+
+borrowck_var_borrow_by_use_place_in_generator =
+    borrow occurs due to use of {$place} in closure in generator
+
+borrowck_var_borrow_by_use_place_in_closure =
+    borrow occurs due to use of {$place} in closure
+
+borrowck_var_borrow_by_use_place =
+    borrow occurs due to use of {$place}

